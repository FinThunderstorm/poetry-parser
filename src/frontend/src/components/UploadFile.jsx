/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { useDispatch } from 'react-redux'
import { styled } from '@mui/material/styles'
import { Button, Grid, Paper, Stack, Typography } from '@mui/material'

import { parsePackages } from '../reducers/packagesSlice'

const UploadFile = () => {
    const dispatch = useDispatch()
    const [selectedFile, setSelectedFile] = useState()
    const navigate = useNavigate()

    const fileSelector = (event) => {
        setSelectedFile(event.target.files[0])
    }
    const handleSubmission = async () => {
        dispatch(parsePackages(selectedFile))
        navigate('/')
    }

    const Input = styled('input')({
        display: 'none',
    })

    return (
        <Grid item xs={12}>
            <Paper elevation={6} sx={{ p: 2 }}>
                <Stack spacing={2} alignItems="center">
                    <Typography variant="button" sx={{ fontSize: '1.5rem' }}>
                        Upload poetry.lock -file
                    </Typography>
                    <Typography>
                        {selectedFile && `Selected file: ${selectedFile.name}`}
                    </Typography>
                    <Stack direction="row" spacing={2}>
                        <label htmlFor="poetry-lock">
                            <Input
                                id="poetry-lock"
                                type="file"
                                name="file"
                                onChange={fileSelector}
                            />
                            <Button variant="outlined" component="span">
                                Select file
                            </Button>
                        </label>

                        <Button
                            variant="contained"
                            type="button"
                            onClick={handleSubmission}
                        >
                            Submit
                        </Button>
                    </Stack>
                </Stack>
            </Paper>
        </Grid>
    )
}

export default UploadFile
