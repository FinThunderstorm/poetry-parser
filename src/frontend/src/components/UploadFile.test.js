/* eslint-disable react/prop-types */
import React from 'react'
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom/extend-expect'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { act } from '@testing-library/react-hooks'
import { Provider } from 'react-redux'
import { MemoryRouter, Outlet, Route, Routes } from 'react-router-dom'
import { configureStore } from '@reduxjs/toolkit'
import axios from 'axios'
import UploadFile from './UploadFile'

import packagesReducer from '../reducers/packagesSlice'

jest.mock('axios') // eslint-disable-line no-undef

describe('UploadFile', () => {
    const UploadFileWrapper = ({ url }) => {
        const store = configureStore({
            reducer: {
                packages: packagesReducer,
            },
        })
        return (
            <MemoryRouter initialEntries={[url]}>
                <Provider store={store}>
                    <Routes>
                        <Route path="/" element={<Outlet />}>
                            <Route path="upload" element={<UploadFile />} />
                        </Route>
                    </Routes>
                </Provider>
            </MemoryRouter>
        )
    }

    test('UploadFile renderes correctly', () => {
        render(<UploadFileWrapper url="/upload" />)

        const title = screen.getByText('UPLOAD POETRY.LOCK -FILE', {
            exact: false,
        })
        expect(title).toBeDefined()

        const tree = renderer
            .create(<UploadFileWrapper url="/upload" />)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })

    test('UploadFile functions correctly', async () => {
        render(<UploadFileWrapper url="/upload" />)

        const fakeFile = new File(['invalid'], 'poetry.lock', {
            type: 'text/plain',
        })

        const input = screen.getByTestId('poetry-lock')
        await act(async () => {
            await userEvent.upload(input, fakeFile)
        })

        expect(input.files[0]).toBe(fakeFile)
        expect(input.files).toHaveLength(1)

        const submitButton = screen.getByText('Submit')
        await act(async () => {
            await userEvent.click(submitButton)
        })

        expect(axios.post).toHaveBeenCalledWith('/parse', fakeFile, {
            headers: { 'Content-Type': 'text/plain' },
        })
    })
})
