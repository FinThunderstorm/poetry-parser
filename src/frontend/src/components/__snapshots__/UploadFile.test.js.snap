// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadFile UploadFile renderes correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 css-1p3yn26-MuiPaper-root"
  >
    <div
      className="css-1y48f0p-MuiStack-root"
    >
      <span
        className="MuiTypography-root MuiTypography-button css-1sqoee-MuiTypography-root"
      >
        Upload poetry.lock -file
      </span>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      />
      <div
        className="css-e53awj-MuiStack-root"
      >
        <label
          htmlFor="poetry-lock"
        >
          <input
            className="css-1hyfx7x"
            id="poetry-lock"
            name="file"
            onChange={[Function]}
            type="file"
          />
          <span
            className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Select file
          </span>
        </label>
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
